aten/src/ATen/native/cpu/BinaryOpsKernel.cpp

```diff
@@ -891,7 +904,14 @@ void heaviside_kernel(TensorIterator& iter) {
 
 template<typename T>
 T copysign(T a, T b) {
+#if (!defined(__aarch64__)) || defined(__clang__) || \
+    (__GNUC__ > 8 || (__GNUC__ == 8 && __GNUC_MINOR__ > 3))
+  // std::copysign gets ICE/Segfaults with gcc 7.5/8 on arm64
+  // (e.g. Jetson), see PyTorch PR #51834
   return std::copysign(a, b);
```

Comment:
Seemed weird to me that copysign was already being used without issue. Looking at the blame, this template comes from #47413 which says it: "avoids internal compiler error exceptions on aarch64 platforms".

So, I'm guessing this doesn't need changed and instead it could be moved somewhere to share it with the CUDA code. Worth a shot at least?

Reply:
Well, so the ICE would only be triggered in specific circumstances.  and so the comment here is probably misleading.
(TBH I wouldn't expect gcc to include completely broken builtins and it seems much more likely that there are corner cases.)
I changed this template because I noticed that tests didn't pass on arm64 and after wild editing it seems better now.
I do think we could take the code that I currently put in CUDAMathCompat.h and put it somewhere shared, but I don't really know where (and I'm only doing this off-and-on and iterating things involving headers has a rather long turnaround time when working on embedded without crosscompiling).

Reply:
The linked PR suggests it's related to calling `std:copysign` with half and BFloat16 types. If that's true then there's no need to avoid `std::copysign` completely.

Reply:
Yeah, moving `copysign` templates solves the issue in the .cpp file for both gcc-7 and gcc-7, let me do it in separate PR. (As well as file a separate issue/gcc bug for that)

Reply:
Ha, that will be much more elegant.

Reply:
Here is the PR that fixes it on CPU (for both gcc-7 and 8)
I wonder if making lambdas GPU only would solve the problem for CUDA

Reply:
So I don't know how this making lambdas GPU only would work, the GPU_LAMBDA makes it `__host__ __device__` as far as I can tell.

Reply:
Yep, GPU_LAMBDA is `__host__ __device__` and there was a valid reason for that (to work around other compiler bugs iirc) which may or may not remain valid. Replacing lambdas with functors with only __device__ `()` functions may work.

Reply:
I think lots of reasons for  `__host__ __device__` are specific to Windows compiler, I wonder if repo-wide effort to replace `__host__ __device__` with `__device__` only on Linux, would speed up compilation and fix those sort of problems.